src/libraries/System.Web.HttpUtility/src/System/Web/Util/HttpEncoder.cs

```diff
@@ -465,6 +438,51 @@ internal static string JavaScriptStringEncode(string? value)
             return expandedBytes;
         }
 
+        private static bool IsExpandingNeeded(ReadOnlySpan<byte> bytes, out int cUnsafe)
+        {
+            cUnsafe = 0;
+
+            int cSpaces = 0;
+            int count = bytes.Length;
+            for (int i = 0; i < count; i++)
```

Comment:
This looks like it could use `SearchValues<byte>` to speed up this initial scan.
For values that don't need encoding, this will be an order of magnitude faster.

Something like
```c#
private static readonly SearchValues<byte> s_urlSafeBytes = SearchValues.Create(
    "!()*-.0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz"u8);

private static bool NeedsEncoding(ReadOnlySpan<byte> bytes, out int cUnsafe)
{
    cUnsafe = 0;

    int i = bytes.IndexOfAnyExcept(s_urlSafeBytes);
    if (i < 0)
    {
        return false;
    }

    foreach (byte b in bytes.Slice(i))
    {
        if (!s_urlSafeBytes.Contains(b) && b != ' ')
        {
            cUnsafe++;
        }
    }

    return true;
}
```

Reply:
Wow - this really makes difference in speed

| Method              | Mean     | Error   | StdDev  | Ratio | Gen0   | Allocated | Alloc Ratio |
|-------------------- |---------:|--------:|--------:|------:|-------:|----------:|------------:|
| UrlEncodeToBytes    | 209.9 ns | 1.21 ns | 1.13 ns |  1.00 | 0.0219 |     184 B |        1.00 |
| UrlEncodeToBytes_PR | 127.0 ns | 1.29 ns | 1.20 ns |  0.61 | 0.0124 |     104 B |        0.57 |

Reply:
SearchValues would be my new favorite toon for optimizations :).
I've used it also to eliminate helper method IsUrlSafeChar.